File: solutions/elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do

  @resistor_color %{
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }

  @doc """
  Calculate the resistance value in ohms from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label(colors) when length(colors) >= 3 do
    [first, second, third] = Enum.take(colors, 3)
    base_value = @resistor_color[first] * 10 + @resistor_color[second]
    multiplier = Integer.pow(10, @resistor_color[third])

    resistance = base_value * multiplier

    cond do
      resistance < 1_000 -> {resistance, :ohms}
      resistance < 1_000_000 -> {div(resistance, 1_000), :kiloohms}
      resistance < 1_000_000_000 -> {div(resistance, 1_000_000), :megaohms}
      true -> {div(resistance, 1_000_000_000), :gigaohms}
    end
  end
end

File: solutions/lua/resistor-color-trio/1/resistor-color-trio.lua
local colors = {
    "black",
    "brown",
    "red",
    "orange",
    "yellow",
    "green",
    "blue",
    "violet",
    "grey",
    "white",
}
local function color_code(color)
    for i, col in ipairs(colors) do
        if col == color then return i - 1 end
    end
end

local function value(c1, c2)
  return color_code(c1) * 10 + color_code(c2)
end

local function shorten_value(v)
  if v == 0 then
    return v, "ohms"
  elseif v % 1000000000 == 0 then
    return v / 1000000000, "gigaohms"
  elseif v % 1000000 == 0 then
    return v / 1000000, "megaohms"
  elseif v % 1000 == 0 then
    return v / 1000, "kiloohms"
  else
    return v, "ohms"
  end
end

return {
  label = function(c1, c2, c3)
    local exp = color_code(c3)
    local v = value(c1, c2) * 10 ^ exp
    return shorten_value(v)
  end
}
